framework/src/org/checkerframework/common/fieldtypeinference/FieldTypeInferenceTransfer.java

```diff
@@ -0,0 +1,112 @@
+ * <tt>org.checkerframework.common.fieldtypeinference.FieldTypeInferenceAnnotatedTypeFactory</tt>.
+ *
+ * @author pbsf
+ */
+
+public class FieldTypeInferenceTransfer extends CFTransfer {
+
+    private final Map<Receiver, AnnotatedTypeMirror> assignedPrivateFields;
+    private final FieldTypeInferenceAnnotatedTypeFactory factory;
+
+    public FieldTypeInferenceTransfer(CFAbstractAnalysis<CFValue, CFStore,
+            CFTransfer> analysis) {
+        super(analysis);
+        assert (analysis.getTypeFactory() instanceof FieldTypeInferenceAnnotatedTypeFactory);
+        factory = (FieldTypeInferenceAnnotatedTypeFactory) analysis.getTypeFactory();
+        assignedPrivateFields = factory.getAssignedPrivateFields();
+    }
+
+    @Override
+    public CFStore initialStore(UnderlyingAST underlyingAST,
+            List<LocalVariableNode> parameters) {
+        CFStore initialStore = super.initialStore(underlyingAST, parameters);
+        for (Receiver r : assignedPrivateFields.keySet()) {
+            for (AnnotationMirror am : assignedPrivateFields.get(r).
+                    getAnnotations()) {
+                initialStore.insertValue(r, am);
+            }
+        }
+        return initialStore;
+    }
+
+    @Override
+    public TransferResult<CFValue, CFStore> visitAssignment(AssignmentNode n,
+            TransferInput<CFValue, CFStore> in) {
+        TransferResult<CFValue, CFStore> result = super.visitAssignment(n, in);
+        assert result instanceof RegularTransferResult;
+        Node rhs = n.getExpression();
+        Receiver expr = FlowExpressions.internalReprOf(analysis.getTypeFactory(),
+                n.getTarget());
+
+        // If this is an assignment to a field of 'this' and this field is
+        // private, then populate the map in the ATF.
+        if (!expr.containsUnknown() && expr instanceof FieldAccess &&
+                ((FieldAccess) expr).getField().getModifiers().
+                contains(Modifier.PRIVATE)) {
+            addAssignedField(expr, factory.getAnnotatedType(rhs.getTree()));
+        }
+        return result;
+    }
+
+    /**
+     * Update the ATM mapped by the parameter <tt>r</tt> in the map
+     * <tt>ClassValAnnotatedTypeFactory.analysis.assignedPrivateFields</tt> to be the LUB
+     * of the parameter <tt>atm</tt> and the current ATM mapped by <tt>r</tt>
+     * in this map.
+     * <p>
+     * If <tt>r</tt> is not a key in the map, it is added to the map
+     * mapping to <tt>atm</tt>.
+     * @param r
+     *      A receiver representing a private field.
+     * @param atm
+     *      An ATM assigned to the private field in <tt>r</tt>.
+     */
+    public void addAssignedField(Receiver r, AnnotatedTypeMirror atm) {
+        Map<Receiver, AnnotatedTypeMirror> assignedPrivateFields = factory.getAssignedPrivateFields();
+        if (assignedPrivateFields.containsKey(r)) {
+            Set<? extends AnnotationMirror> lub = analysis.getTypeFactory().
```

Comment:
Should we instead be using the LUB for AnnotatedTypeMirrors?  The RHS of assignments may be below the TypeMirror so we would have to convert them to the declaration BUT since, lub rules differ depending on whether or not the type being lubbed is a type variable, we might need to consider the underlying type.

Please observe the following type parameter and field declaration:
<@Regex(3) T extends @Regex(1) Object>   // this has a type  <T extends @Regex(1) super @Regex(3) Void>
@Regex(0) T t0;                                               // this has a type <T extends @Regex(0) super @Regex(0) Void>
T nakedT;
T nakedT2;

Given these assignments:
t0 = nakedT;
t0 = nakedT2;

If t0 should be refined to the lub of nakedT and nakedT2, then it's type should be:
 < T extends @Regex(1) Object super @Regex(3) T> 

which is actually BELOW @Regex(0) T

So we can't just get the primary annotations and take the lub.  Otherwise, we would only have 1 primary annotation @Regex(0).  And this would be an overly conservative estimate.  Note, this can also work the other way where the estimate is also overly optimistic and therefore unsound.